larpandoracontent/LArTwoDReco/LArClusterAssociation/CrossGapsExtensionAlgorithm.cc

```diff
@@ -51,72 +51,90 @@ void CrossGapsExtensionAlgorithm::GetListOfCleanClusters(const ClusterList *cons
 
 void CrossGapsExtensionAlgorithm::FillClusterAssociationMatrix(const ClusterVector &clusterVector, ClusterAssociationMatrix &clusterAssociationMatrix) const
 {
-    // Build lists of pointing clusters in proximity to gaps
-    LArPointingClusterList innerPointingClusterList, outerPointingClusterList;
-    this->BuildPointingClusterList(clusterVector, innerPointingClusterList, outerPointingClusterList);
+    // Build and loop over list of pointing clusters
+    LArPointingClusterList pointingClusterList;
+    this->BuildPointingClusterList(clusterVector, pointingClusterList);
 
-    // Form associations between pairs of pointing clusters
-    for (const LArPointingCluster &pointingClusterInner : innerPointingClusterList)
+    for (LArPointingClusterList::const_iterator iter1 = pointingClusterList.begin(), iterEnd1 = pointingClusterList.end();
+        iter1 != iterEnd1; ++iter1)
     {
-        const LArPointingCluster::Vertex &pointingVertexInner(pointingClusterInner.GetInnerVertex());
-        const float zInner(pointingVertexInner.GetPosition().GetZ());
+        const LArPointingCluster &pointingCluster1 = *iter1;
+        const Cluster *const pCluster1(pointingCluster1.GetCluster());
 
-        for (const LArPointingCluster &pointingClusterOuter : outerPointingClusterList)
+        for (LArPointingClusterList::const_iterator iter2 = iter1, iterEnd2 = pointingClusterList.end(); iter2 != iterEnd2; ++iter2)
         {
-            const LArPointingCluster::Vertex &pointingVertexOuter(pointingClusterOuter.GetOuterVertex());
-            const float zOuter(pointingVertexOuter.GetPosition().GetZ());
+            const LArPointingCluster &pointingCluster2 = *iter2;
+            const Cluster *const pCluster2(pointingCluster2.GetCluster());
 
-            if (!this->IsAcrossGap(zOuter, zInner, LArClusterHelper::GetClusterHitType(pointingClusterInner.GetCluster())))
+            if (pCluster1 == pCluster2)
                 continue;
 
-            if (!this->IsAssociated(pointingVertexInner, pointingVertexOuter))
+            // Get hit types and check they're the same
+            const HitType hitType1(LArClusterHelper::GetClusterHitType(pCluster1));
+            const HitType hitType2(LArClusterHelper::GetClusterHitType(pCluster2));
+
+            if (hitType1 != hitType2)
+                throw StatusCodeException(STATUS_CODE_FAILURE);
+
+            // Get closest pair of vertices from pointing clusters
+            LArPointingCluster::Vertex closestVertex1, closestVertex2;
+
+            try
+            {
+                LArPointingClusterHelper::GetClosestVertices(pointingCluster1, pointingCluster2, closestVertex1, closestVertex2);
+            }
+            catch (StatusCodeException &)
```

Comment:
`catch (const StatusCodeException &)`

There are rather too many exceptions being raised, and caught as a matter of routine in LArContent...